content/client-lib-development-guide/features.textile

```diff
@@ -468,7 +468,7 @@ h3(#realtime-presence). Presence
 * @(RTP1)@ When a channel @ATTACHED@ @ProtocolMessage@ is received, the @ProtocolMessage@ may contain a @HAS_PRESENCE@ bit flag indicating that there are currently members present on the channel, see "TR3":#TR3 . If the flag is 1, the server will shortly perform a @SYNC@ operation as described in "RTP18":#RTP18 . If that flag is 0 or there is no @flags@ field, the presence map should be considered in sync immediately with no members present on the channel
 * @(RTP2)@ A @PresenceMap@ should be used to maintain a list of members present on a channel. Broadly, this is is a map of "memberKeys":#TP3h to presence messages, all with @PRESENT@ actions (during a sync there may also be ones with an @ABSENT@ action, see "RTP2f":#RTP2f ). Consult the "the Java implementation":https://github.com/ably/ably-java/blob/master/lib/src/main/java/io/ably/lib/realtime/Presence.java#L578-L723 for details.
 ** @(RTP2a)@ All incoming presence messages must be compared for newness with the matching member already in the @PresenceMap@, if one exists, where "matching" means they share the same @memberKey@ (or equivalently, they share both @connectionId@ and @clientId@)
-** @(RTP2b)@ 'Comparing for newness' involves: If the @id@ of an incoming presence message corresponds to its @connectionId@ (defined as the @connectionId@ being an initial substring of the @id@), then the messages are compared lexicographically by @id@. If not, they are compared by timestamp
+** @(RTP2b)@ 'Comparing for newness' involves: If the @id@ of both presence message corresponds to the @connectionId@ (defined as the @connectionId@ being an initial substring of the @id@), then the messages are compared lexicographically by @id@. If not, they are compared by timestamp
```

Comment:
Oh. Yeah, you're right. Bugger.

Which means that to do the comparison correctly, we have to split the `id` on `-` and `:` to get the msgSerial and index, pass both through parseInt, and compare first on msgSerial, then index. Though actually we can't even do a simple split because `-` is a valid character in a connectionId as well as the connectionId-msgserial separator, but we could do it with a regex with capture groups - something like `[\w-_]*-(\d+):(\d+)`. Probably put that in the spec, and make sure we never change the format of the `id`.

Is it just me, or is this starting to seem like a crazy thing for us to require the client lib to do for every incoming presence message? Surely we can think of a better way.

@paddybyers how about if realtime adds a logical clock field with every presence message? i.e. a number that defines an ordering between all presence messages that share a connectionId. Effectively an order-preserving injection from {msgSerial, index} to the naturals. Then the client libs could just compare by integer.